sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala

```diff
@@ -1174,4 +1174,33 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
         }
       }
   }
+
+  test("SPARK-21024 CSV parser mode controls parser exceptions") {
+    withTempPath { path =>
+      Seq("0,1", "0,1,2,3").toDF().write.text(path.getAbsolutePath)
+
+      Seq(false).foreach { wholeFile =>
+        val msg = intercept[SparkException] {
+          spark.read.format("csv")
+            .schema("a INT, b INT")
+            .option("maxColumns", "2")
+            .option("mode", "FAILFAST")
+            .option("wholeFile", wholeFile)
+            .load(path.getAbsolutePath)
+            .collect
+        }.getMessage
+        assert(msg.contains("Number of columns processed may have exceeded limit of 2 columns."))
+
+        val columnNameOfCorruptRecord = "_unparsed"
+        val df = spark.read.format("csv")
+          .schema(s"a INT, b INT, $columnNameOfCorruptRecord STRING")
+          .option("maxColumns", "2")
+          .option("mode", "PERMISSIVE")
+          .option("columnNameOfCorruptRecord", columnNameOfCorruptRecord)
+          .option("wholeFile", wholeFile)
+          .load(path.getAbsolutePath)
+        checkAnswer(df, Row(0, 1, null) :: Row(null, null, "0,1,2,") :: Nil)
```

Comment:
@HyukjinKwon weird behaviour..., when we set `maxColumns` in a Univocity parser, it seems `currentParsedContent` returns the (`maxColumns` + 1) elements in inputs.